lib/afinter.c

```diff
@@ -635,3 +643,107 @@ afinter_get_metrics(void)
 {
   return metrics;
 }
+
+/****************************************************************************
+ * Live collection of internal log messages.
+ ****************************************************************************/
+
+AFInterLive
+afinter_start_live_collection(void)
+{
+  AFInterSource *self;
+
+  g_mutex_lock(&internal_msg_lock);
+
+  if (is_live_collection)
+    return AFINTER_LIVE_COLLECTION_RUNNING;
+
+  if (current_internal_source != NULL)
+    return AFINTER_INTERNAL_SRC_PRESENT;
+
+  self = g_new0(AFInterSource, 1);
+  AFInterSourceOptions *options = g_new0(AFInterSourceOptions, 1);
+
+  afinter_source_options_defaults(options);
+  self->options = options;
+  stats_counter_set(metrics.queue_capacity, self->options->queue_capacity);
+
+  current_internal_source = self;
+  is_live_collection = TRUE;
+
+  g_mutex_unlock(&internal_msg_lock);
+
+  return AFINTER_LIVE_COLLECTION_STARTED;
+}
+
+void
+afinter_reset_live_collection(void)
+{
+  afinter_stop_live_collection();
+
+  if (is_live_collection)
+    {
+      if (internal_msg_queue)
+        _release_internal_msg_queue();
+    }
+}
+
+void
+afinter_stop_live_collection(void)
+{
+  g_mutex_lock(&internal_msg_lock);
+  if (is_live_collection)
+    {
+      if (current_internal_source != NULL)
+        {
+          AFInterSource *self = current_internal_source;
+
+          current_internal_source = NULL;
+
+          is_live_collection = FALSE;
+          g_free(self);
+        }
+    }
+  g_mutex_unlock(&internal_msg_lock);
+}
+
+void
+afinter_get_size_of_internal_logs(GString *result)
+{
+  g_string_append_printf(result, "Size of internal logs: %d", g_queue_get_length(internal_msg_queue));
+}
+
+void
+afinter_get_collected_messages(GString *result)
+{
+  LogMessage *log_msg;
+  GString *msg;
+  LogTemplate *template = NULL;
+
+  g_mutex_lock(&internal_msg_lock);
```

Comment:
my main concerns here are
- the queue can grow to a huge size here that can eat up all the memories if a huge number of messages are coming
- it also means a doubled memory usage because of the copied content in the result response string
- locking the queue here, during the whole content processing, means a lock on the sender side too, as I can see on the main thread that will stop syslog-ng message processing for this time, yes, I know, it is called only now after the log processing stopped, but as this function publicly accessible, nothing prevents it to be called from anywhere else.
I suggest making this static local only and calling it from the stop directly, internally.

Reply:
@HofiOne : okay, good one.